### PR TITLE
Changed handling of websocket connection and reconnection

### PR DESCRIPTION
**Pull request by ethankan**
This makes sure we don't schedule reconnection multiple times when websocket pings fail.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -138,82 +89,170 @@ ModeDevice.prototype.setUpApiHttp = function() {
...
   }
 
-  // Only if less than 60 sec, we increment waiting time according to Fibonacci numbers.
-  if (this.retryWait < 60) {
-    var fib = this.retryWaitFib;
-    this.retryWaitFib = this.retryWait;
-    this.retryWait += fib;
+  if (this._wsReconnectDelay < 60) {
+    // exponential backoff
+    this._wsReconnectDelay = Math.pow(2, this._wsReconnectAttempts);
```
Comment (honteng) [inline index.js:210]: just curious why we change from Fib to pow algorithm? Is it better?

Comment (honteng) [reply index.js:210]: Because pow is earlier to reach to the limit 60?

File: index.js [hunk 2]
```diff
@@ -138,82 +89,170 @@ ModeDevice.prototype.setUpApiHttp = function() {
...
+
+  this._wsReconnectAttempts++;
+  debuglog('Retrying websocket connection in %d seconds (attempt #%d)', this._wsReconnectDelay, this._wsReconnectAttempts);
+
+  // Make sure we never double-schedule reconnection.
+  if (this._wsReconnectTimer) {
+    clearTimeout(this._wsReconnectTimer);
+  }
+
+  this._wsReconnectTimer = setTimeout(this._wsConnect.bind(this), this._wsReconnectDelay * 1000);
```
Comment (honteng) [inline index.js:221]: ah, it looks pretty same reason as we encountered in DDP bug?

File: index.js [hunk 3]
```diff
@@ -138,82 +89,170 @@ ModeDevice.prototype.setUpApiHttp = function() {
...
       "Authorization": 'ModeCloud ' + this.token
     }
   });
-  this.websocket.on('error', this.errorCallback.bind(this));
-  this.websocket.on('close', this.closeCallback.bind(this));
-  this.websocket.on('open', this.openCallback.bind(this));
-  this.websocket.on('message', commandHandler.bind(this));
-  this.websocket.on('pong', this.pongCallback.bind(this));
+
+  this._websocket.on('error', this._wsHandleError.bind(this));
```
Comment (honteng) [inline index.js:193]: So, from this version, we don't give the user any freedom to access to the websocket error handling?

Comment (ethankan) [reply index.js:193]: The reconnection logic is built into the error and close handler so they should not be overridden, or the whole thing may break. In a future version, we can probably turn this into an emitter and emit 'connected' and 'disconnected' events to notify the client of a change in the websocket state. But those should be informational only, and not meant to client to override the behavior of the library.

File: index.js (hunk 2 above)
Comment (ethankan) [reply index.js:221]: Actually, in both cases, the problem was caused by the fact that when a ping fails, we were calling close() *and* reconnect(). Then the onClose handler will also try to call reconnect(). And we ended up with two "threads" competing with each other.

**Pull request merged**